UnitySDK/Assets/ML-Agents/Scripts/DecisionAutoRequester.cs

```diff
@@ -0,0 +1,39 @@
+﻿using UnityEngine;
+
+namespace MLAgents
+{
+    [RequireComponent(typeof(Agent))]
+    public class DecisionAutoRequester : MonoBehaviour
+    {
+
+        public int DecisionPeriod = 1;
+        public bool RepeatAction = true;
+        private Agent m_Agent;
+        private Academy m_Academy;
+
+        public void OnEnable()
+        {
+            DecisionPeriod = Mathf.Max(DecisionPeriod, 1);
+            m_Agent = gameObject.GetComponent<Agent>();
+            m_Academy = GameObject.FindObjectOfType<Academy>();
+            m_Academy.AgentSetStatus += SetStatus;
+        }
+
+        void OnDisable()
+        {
+            m_Academy.AgentSetStatus -= SetStatus;
+        }
+
+        void SetStatus(bool academyMaxStep, bool academyDone, int academyStepCounter)
```

Comment:
What if we 
1) Keep AgentParameters.onDemandDecision and numberOfActionsBetweenDecisions
2) Move this logic to the agent, and register it with the academy if onDemandDecision is false
That lets you keep most of this cleanup, but avoids the extra work of users adding these DecisionAutoRequesters.

(Might not quite work if AgentParameters fields are mutable at runtime, but I think it's still fixable)

Reply:
@chriselion , is it really that big an issue to have to add a DecisionAutoRequester to the Agent? I am trying to remove functionality from the Agent with this PR.  @ervteng any opinion on this?

Reply:
Very rough sketch here https://github.com/Unity-Technologies/ml-agents/pull/2581